### PR TITLE
Add new endpoint for: Search or create for user

### PR DESCRIPTION
**Pull request by ericnormand**
When a client queries for existing electorates, the electorate created
from user create events might not have finished. So **this new endpoint**
tries to create them if none exist yet. It still might return an empty
set if there are no upcoming elections. In the vast majority of cases,
there will be either one or zero upcoming elections, and this endpoint
is made to work well with those cases. In the small likelihood that
there are multiple upcoming elections, it is possible that one
electorate will be created and the others have not finished yet, and
that single created one will be returned. We weighed the cost of this
problem with the complexity of solving it (essentially equivalent to the
cache invalidation problem) and decided it was not worth the cost to
solve.

So there is now one endpoint for *search or attempt to create if none are found*. The *search* branch
reuses the same codepaths as the old search, and the *attempt to
create branch* reuses the same codepaths as the regular *create
electorates from user create event*.

**Status**: Electorates are correctly returned with expected data.

### CODE REVIEW

File: src/electorate_works/handlers.clj [hunk 1]
```diff
@@ -194,13 +194,22 @@
...
                  (assoc-when :registration-authority @registration-authority)))))))))
 
+(defn create-and-broadcast-electorates-from-user
+  [user search-elections search-election-authorities timeout]
+  (let [electorates (create-electorates-from-user search-elections
+                                                  search-election-authorities
+                                                  user
+                                                  timeout)]
+    (doseq [electorate electorates]
+      (create-and-broadcast-electorate electorate))
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: This makes it seem like we're creating the electorates twice.

File: src/electorate_works/handlers.clj [hunk 2]
```diff
@@ -194,13 +194,22 @@
...
 (defn handle-user-create-event
   [user]
-  (doseq [electorate (create-electorates-from-user search-elections
-                                                   search-election-authorities
-                                                   user
-                                                   5000)]
-    (create-and-broadcast-electorate electorate)))
+  (create-and-broadcast-electorates-from-user user search-elections
+                                              search-election-authorities
+                                              500)
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: Should this still be 5 seconds? 0.5 seconds seems pretty short.

File: src/electorate_works/handlers.clj [hunk 3]
```diff
@@ -309,3 +318,41 @@
...
+           :electorates (->> (create-and-broadcast-electorates-from-user
+                              user
+                              search-elections-fn
+                              search-election-authorities-fn
+                              5000)
+                             p
+                             (filter future-electorate?)
+                             p
+                             set
+                             )})
```
Comment (cap10morgan) [inline src/electorate_works/handlers.clj]: Should we remove the calls to `p` now (and `defn p` itself)?

File: test/electorate_works/handlers_test.clj [hunk 4]
```diff
@@ -426,6 +426,241 @@
                       (get-in % [:election :description]))
                   (:electorates response)))))))
 
+(defn singletonius [val]
+  (let [c (async/chan)]
```
Comment (cap10morgan) [inline test/electorate_works/handlers_test.clj]: Would love to see a docstring on this guy.

File: test/electorate_works/handlers_test.clj [hunk 5]
```diff
@@ -426,6 +426,241 @@
...
+                                                           {:user-id (:id user)})]
+        (is (= :ok (:status response)))
+        (is (= 2 (count (:electorates response))))
+        (is (some #(= "Worst Hair Piece Election"
+                      (get-in % [:election :description]))
+                  (:electorates response)))
+        (is (some #(= "Other Election"
+                      (get-in % [:election :description]))
+                  (:electorates response)))))))
+
```
Comment (cap10morgan) [inline test/electorate_works/handlers_test.clj]: Lots of tests, awesome!

File: README.md [hunk 6]
```diff
@@ -59,6 +59,39 @@ A successful response is as follows.
...
+{:user-id #uuid "a099df06-9ce3-4537-8fbc-2e48f597a8fd"}
+```
+
+A successful response returns all known electorates for that user:
+
+```
+{:status :ok
+ :electorates #{{...} ...}}
+```
+
```
Comment (cap10morgan) [inline README.md]: Did we resolve the question of whether search was still needed?

Comment (ericnormand) [inline README.md]: My searches indicate that nothing is using it. I recommend we remove it.

**Comment by ericnormand** [comment]
Can I get a LGTM?

**Comment by cap10morgan** [comment]
LGTM

**Pull request merged**